azure-pipelines.yml

```diff
@@ -3,135 +3,258 @@
+    env:
+      SPARK_HOME: $(Build.BinariesDirectory)\spark-2.3.2-bin-hadoop2.7
+      HADOOP_HOME: $(Build.BinariesDirectory)\hadoop
+      DotnetWorkerPath: $(Build.ArtifactStagingDirectory)\Microsoft.Spark.Worker\netcoreapp2.1\win-x64
+
+  - task: DotNetCoreCLI@2
+    displayName: 'E2E tests for Spark 2.3.3'
+    inputs:
+      command: test
+      projects: '**/Microsoft.Spark.E2ETest/*.csproj'
+      arguments: '--configuration $(buildConfiguration) /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura'
+    env:
+      SPARK_HOME: $(Build.BinariesDirectory)\spark-2.3.3-bin-hadoop2.7
+      HADOOP_HOME: $(Build.BinariesDirectory)\hadoop
+      DotnetWorkerPath: $(Build.ArtifactStagingDirectory)\Microsoft.Spark.Worker\netcoreapp2.1\win-x64
+
+  - task: DotNetCoreCLI@2
+    displayName: 'E2E tests for Spark 2.4.0'
+    inputs:
+      command: test
+      projects: '**/Microsoft.Spark.E2ETest/*.csproj'
+      arguments: '--configuration $(buildConfiguration) /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura'
+    env:
+      SPARK_HOME: $(Build.BinariesDirectory)\spark-2.4.0-bin-hadoop2.7
+      HADOOP_HOME: $(Build.BinariesDirectory)\hadoop
+      DotnetWorkerPath: $(Build.ArtifactStagingDirectory)\Microsoft.Spark.Worker\netcoreapp2.1\win-x64
+
+  - task: DotNetCoreCLI@2
+    displayName: 'E2E tests for Spark 2.4.1'
+    inputs:
+      command: test
+      projects: '**/Microsoft.Spark.E2ETest/*.csproj'
+      arguments: '--configuration $(buildConfiguration) /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura'
+    env:
+      SPARK_HOME: $(Build.BinariesDirectory)\spark-2.4.1-bin-hadoop2.7
+      HADOOP_HOME: $(Build.BinariesDirectory)\hadoop
+      DotnetWorkerPath: $(Build.ArtifactStagingDirectory)\Microsoft.Spark.Worker\netcoreapp2.1\win-x64
+  
+  - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+    - task: CopyFiles@2
+      displayName: Stage .NET artifacts
+      inputs:
+        sourceFolder: $(Build.SourcesDirectory)/src/csharp/Microsoft.Spark/bin/$(buildConfiguration)
+        contents: |
+          **/*.dll
+          **/*.pdb
+        targetFolder: $(Build.ArtifactStagingDirectory)/BuildArtifacts/src/csharp/Microsoft.Spark/bin/$(buildConfiguration)
+
+    - task: CopyFiles@2
+      displayName: Stage scala artifacts
+      inputs:
+        sourceFolder: $(Build.SourcesDirectory)/src/scala
+        contents: '**/*.jar'
+        targetFolder: $(Build.ArtifactStagingDirectory)/BuildArtifacts/src/scala
+    
+    - task: PublishBuildArtifacts@1
+      inputs:
+        pathtoPublish: '$(Build.ArtifactStagingDirectory)'
+        artifactName:  Microsoft.Spark.Binaries
+
+- ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+  - job: Package
+    dependsOn:
+      - Build
+    displayName: Build packages
+    pool:
+      name: NetCoreInternal-Int-Pool
```

Comment:
Why not just run the whole build on `NetCoreInternal-Int-Pool` and collapse this into a single job?  That way you don't have to publish the artifacts and then pull them down. It could be simpler to do it all in 1 job.

Reply:
Ah, I see we weren't before. So maybe this is a question for @imback82 and @rapoth

---
In reply to: [278199899](https://github.com/dotnet/spark/pull/6#discussion_r278199899) [](ancestors = 278199899)

Reply:
That was the initial approach but the machines in `NetCoreInternal-Int-Pool` don't have Maven in it and we need Maven for the scala sources.

Reply:
Just an FYI - that might be a good comment to have in the .yml file.

Also - can we fix `NetCoreInternal-Int-Pool` to have it?

Reply:
I talked to @chcosta yesterday and he recommended this approach, but yes I agree, maybe the `NetCoreInternal-Int-Pool` should have it. cc: @mattgal is this possible? 

I'll add a comment in the meantime

Reply:
@safern the general guidance is to bootstrap anything in builds which can be done in a non-stateful-to-machine way;  I'll do some reading on Maven and if it makes sense we can add it to these machines (though it will take a bit to roll out)